src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -790,6 +794,38 @@ void ArchiveBuilder::make_klasses_shareable() {
   log_info(cds)("    obj array classes  = %5d", num_obj_array_klasses);
   log_info(cds)("    type array classes = %5d", num_type_array_klasses);
   log_info(cds)("               symbols = %5d", _symbols->length());
+
+  DynamicArchive::make_array_klasses_shareable();
+}
+
+void ArchiveBuilder::gather_array_klasses() {
+  for (int i = 0; i < klasses()->length(); i++) {
+    if (klasses()->at(i)->is_objArray_klass()) {
+      ObjArrayKlass* oak = ObjArrayKlass::cast(klasses()->at(i));
+      Klass* elem = oak->element_klass();
+      if (MetaspaceShared::is_shared_static(elem)) {
+        // Only capture the array klass whose element_klass is in the static archive.
+        // During run time, setup (see DynamicArchive::setup_array_klasses()) is needed
+        // so that the element_klass can find its array klasses from the dynamic archive.
+        ArrayKlass* ak = ArrayKlass::cast(klasses()->at(i));
+        DynamicArchive::append_array_klass(ak);
+      } else {
+        // The element_klass and its array klasses are in the same archive.
+        if (oak->dimension() > 1) {
+          assert(ArrayKlass::cast(elem)->higher_dimension() == oak, "must be");
+        } else {
+          assert(InstanceKlass::cast(elem)->array_klasses() == oak, "must be");
+        }
+      }
+    }
+  }
+  log_debug(cds)("Total array klasses gathered for dynamic archive: %d", DynamicArchive::num_array_klasses());
+}
```

Comment:
This function can be put inside DynamicArchiveBuilder, since it's not used by the static dump.

Reply:
Fixed